Statistics

florianl opened the pull request:
Signed-off-by: Florian Lehner <dev@der-flo.net>

    BPF can provide kernel internal statistics about how often a program was
    called and its runtime in ns.
    This PR adds the functionallity to enable the collection of these
    statistics and provides a way to fetch them.

nathanjsweet commented on internal/syscall.go:
```diff
@@ -91,6 +91,19 @@ func BPFProgDetach(attr *BPFProgDetachAttr) error {
+}
+
+func BPFEnableStats(attr *BPFEnableStatsAttr) error {
+	_, err := BPF(BPF_ENABLE_STATS, unsafe.Pointer(attr), unsafe.Sizeof(*attr))
```
I would like to see a feature test for this function. [See objectNames](https://github.com/cilium/ebpf/blob/master/syscalls.go#L399) as an example.

nathanjsweet reviewed (changes requested):
Could you flesh out the commit message a bit to explain what this is. I would also like to see an integration test as well.

florianl commented:
Sorry about being short worded in the commit message.
I've extended the commit message and added the requested feature test.

nathanjsweet commented on info.go:
```diff
@@ -182,3 +182,19 @@ func scanFdInfoReader(r io.Reader, fields map[string]interface{}) error {
+	return internal.BPFEnableStats(&attr)
+}
+
+var canEnableStats = internal.FeatureTest("can enable stats", "5.8", func() error {
```
Sorry. I didn't realize that there was no way to disable runtime stats after they were enabled. I guess this feature test is actually kind of useless. I recommend removing it.

lmb replied:
Agree. Is it possible to return a wrapped ErrNotSupported based on the return value from BPFEnableStats?

For future reference, you need to return ErrNotSupported from feature tests in some form, otherwise there isn't much point to it.

florianl replied:
i have removed the feature test again.

florianl commented on info.go:
```diff
@@ -182,3 +182,19 @@ func scanFdInfoReader(r io.Reader, fields map[string]interface{}) error {
+// and run counts.
+//
+// Requires at least 5.8.
+func EnableStats() error {
```
At the moment only `BPF_STATS_RUN_TIME` is supported to be set as value in the attribute. Maybe (I don't know) the kernel will provide more `StatsTyp` at some point.
Do we want to change the API then or ask now the user to provide an argument to `EnableStats()`?

lmb replied:
I'd go with an extra parameter since the kernel API encourages extension. I wonder whether it should be `(which int)` or `(which StatsType)`.

florianl replied:
I added the extra argument for the function.

lmb commented on internal/syscall.go:
```diff
@@ -91,6 +91,19 @@ func BPFProgDetach(attr *BPFProgDetachAttr) error {
 }
 
+const (
+	BPF_STATS_RUN_TIME = iota
```
Move this to `internal/unix` instead? golang.org/x/sys/unix has the variable

lmb commented on info.go:
```diff
@@ -182,3 +182,19 @@ func scanFdInfoReader(r io.Reader, fields map[string]interface{}) error {
 }
+
+// EnableStats starts the measuring of the runtime (in ns)
+// and run counts.
```
Document that this can't be turned off?

lmb commented on prog.go:
```diff
@@ -613,6 +613,24 @@ func (p *Program) ID() (ProgramID, error) {
+}
+
+// Stats returns measured statistics of a the program.
+func (p *Program) Stats() (ProgStats, error) {
```
This should be on ProgramInfo, with a getter similar to `ProgramInfo.ID()`. Also document that this returns zero (?) values without EnableStats.

florianl replied:
I decided against `func (i *ProgramInfo) Stats() (ProgramStats, error) ` as it requires more syscalls to get the same information.
E.g. one has to get the FD from `ProgramInfo` first before `bpfGetProgInfoByFD` can be used. And storing FD in `ProgramInfo` feels not correct. what do you think?

lmb replied:
Take a look at newProgramInfoFromFd, which is called by Program.Info. There you can pull out the necessary info for ProgramStats and store it in an unexported field in ProgramInfo. The getter then allows the user to retrieve the `ProgramStats` while also telling them whether the feature is available in the first place (on old kernels we can't populate the info). All of this is still only a single syscall.

lmb commented on prog.go:
```diff
@@ -613,6 +613,24 @@ func (p *Program) ID() (ProgramID, error) {
 }
 
+// ProgStats holds statistics of a program.
+type ProgStats struct {
```
```suggestion
type ProgramStats struct {
```

ti-mo commented on internal/syscall.go:
```diff
@@ -91,6 +91,15 @@ func BPFProgDetach(attr *BPFProgDetachAttr) error {
 }
 
+type BPFEnableStatsAttr struct {
+	StatsTyp uint32
```
This can be `StatsType`

nathanjsweet reviewed (approved):
I'm happy, but I will defer to you and @lmb working out your differences on where the Stats function should live. I see the merits of both arguments.

ti-mo commented on info.go:
```diff
@@ -182,3 +182,18 @@ func scanFdInfoReader(r io.Reader, fields map[string]interface{}) error {
 	return nil
 }
+
+// EnableStats starts the measuring of the runtime (in ns)
```
This reads a bit vaguely to me. I'm not familiar with this particular feature, so I immediately had the following questions:
- Which programs get their runtime stats collected? Surely not all BPF programs on the machine, that's why this syscall was introduced.
- How can the stats be queried?
- What am I supposed to pass to `which`?

florianl replied:
> Which programs get their runtime stats collected? Surely not all BPF programs on the machine, that's why this syscall was introduced.

As far as I can tell, at the moment there is only the option to enable the colleciton of statistics. If enabled statistics for all executing eBPF programs are collected.
I can imagine, this statistic collection will be extendend and maybe more specific at some point in the future. 

> What am I supposed to pass to which?
At the moment, there is only a single valid value for `which` - `BPF_STATS_RUN_TIME`. `which` was introduced because of discussions in https://github.com/cilium/ebpf/pull/213#discussion_r574313592. 
As `BPF_STATS_RUN_TIME` is provided by golang.org/x/sys/unix I did not add an additional wrapper.

ti-mo commented on prog.go:
```diff
@@ -613,6 +613,26 @@ func (p *Program) ID() (ProgramID, error) {
 
+// ProgramStats holds statistics of a program.
+//
+// Values can be 0 if collecting of statistics is not enabled.
```
This semantic should be documented on `(*Program).Stats()` instead.

ti-mo commented on prog.go:
```diff
@@ -613,6 +613,26 @@ func (p *Program) ID() (ProgramID, error) {
+// ProgramStats holds statistics of a program.
+//
+// Values can be 0 if collecting of statistics is not enabled.
+type ProgramStats struct {
```
Since this is an exported struct, IMO these fields should be properly capitalized and expanded as is the convention in Go: `RuntimeNS` and `RunCount`.

ti-mo commented on prog.go:
```diff
@@ -613,6 +613,26 @@ func (p *Program) ID() (ProgramID, error) {
 	return ProgramID(info.id), nil
 }
 
+// ProgramStats holds statistics of a program.
```
Could you elaborate on the meaning of the values in the struct? e.g. is `RuntimeNS` the total accumulated runtime of the program? Or the latest invocation? Or the average?

ti-mo reviewed (changes requested):
Thanks! I would like to see at least a test or an example of how to use this, though.

lmb commented:
I just found out that BPF_STATS_ENABLE returns a file descriptor, which disables statistics collection when it's closed!

lmb reviewed (changes requested):
* Please take a look at my suggestion for ProgramInfo.Stats(): no extra syscall and we avoid cluttering Program.
* EnableStats needs to return an object that wraps an `internal.FD`, not sure what to call that, `StatsHandle`? 
* Please add a test that enables stats, runs a prog, checks the count, disables the stats.

ti-mo commented:
> I just found out that BPF_STATS_ENABLE returns a file descriptor, which disables statistics collection when it's closed!

I also suspected this from reading the kernel patches, this seemed the whole point of the syscall. Have you been able to play around with this? Does it only affect BPF programs the current process holds a reference to? What about pinned programs?

lmb commented:
I think it affects all programs, and is reetrant: multiple calls to enable stats can hold an fd at the same time, and collection is only stopped when the last one is closed. pinned programs are no different than regular programs, so those will get stats collected as well.

florianl commented:
Thanks for your feedback :pray: I didn't realize it was possible to disable the collecting of statistics.

I have updated the PR to use `ProgramInfo.Stats()`, added a wrapper function similar to `[context.CancelFunc](https://pkg.go.dev/context#CancelFunc)` to disable the collection of statistics.
And I added a test that fetches statistics before and after `EnableStats` and after disabling the statistics.

lmb commented on info.go:
```diff
@@ -127,6 +130,28 @@ func (pi *ProgramInfo) ID() (ProgramID, bool) {
+	if err != nil {
+		return ProgramStats{}, err
+	}
+	return ProgramStats{
```
Why not create this when doing the bpfGetProgInfoByFD here? https://github.com/cilium/ebpf/blob/7a313e22540feeda41df7b82432ab06c9ee50b78/info.go#L84
Instead of adding `*internal.FD` to `ProgramInfo` add `stats *ProgramStats`. Keeping the fd around like that without a `Close()` function is basically a leak.

This function can then be `Stats() (ProgramStats, bool)` with the following body:

```
if pi.stats != nil { return *pi.stats, true }
return ProgramStats{}, false
```

* A single syscall
* It's clear that ProgramStats may be absent (like on older kernels)
* No dangling fds

lmb replied:
Something I don't have a strong opinion on, but that is worth thinking about: is the `ProgramStats` type necessary? We could  have two getters on `ProgramInfo` instead: `ProgramInfo.Runtime() (foo, bool)` and `ProgramInfo.RunCount(foo, bool)`.

florianl replied:
Ok - I think I get your point.

The idea I had in mind was to be able to repeatly call `Stats()` and get the current information directly. While I understand your approach more like get `ProgramInfo` and from this get the statistics. With your approch getters are more usefull, I think.

florianl replied:
I've chosen `ProgramInfo.RunCount()(uint64)` over `ProgramInfo.RunCount()(uint64, bool)`, as we can not decide from the available data, if stats are enabled or not. e.g. `0` is a valid return if the program was never called.
If one can not call `EnableStats()` without error, then the kernel doesn't provide this information. So it is more an indirect check via the enable function.

lmb commented on info.go:
```diff
@@ -182,3 +207,24 @@ func scanFdInfoReader(r io.Reader, fields map[string]interface{}) error {
+// Collecting statistics can have an impact on the performance.
+//
+// Requires at least 5.8.
+func EnableStats(which uint32) (DisableStatsFunc, error) {
```
```suggestion
func EnableStats(which uint32) (io.Closer, error) {
```

IIRC internal.FD should already implement `Closer`. That way we have one less exported type, and returning an interface is more flexible than a function. If the API sprouts some additional usage of the FD we can extend the returned interface.

lmb reviewed (changes requested):
Tests LGTM, thanks! I think there is still some miscommunication around ProgramInfo, I left another comment, hopefully that is clearer. Let me know if it makes sense.

lmb commented on info.go:
```diff
@@ -127,6 +130,28 @@ func (pi *ProgramInfo) ID() (ProgramID, bool) {
+// ProgramStats holds statistics of a program.
+type ProgramStats struct {
+	// Total accumulated runtime of the program ins ns.
+	RuntimeNS uint64
```
```suggestion
	Runtime time.Duration
```

This is probably nicer?
